src/main/webapp/app/exercises/programming/shared/instructions-render/extensions/programming-exercise-plant-uml.extension.ts

```diff
@@ -66,19 +67,22 @@ export class ProgrammingExercisePlantUmlExtensionWrapper implements ArtemisShowd
                 const plantUmlContainer = `<div class="mb-4" id="plantUml-${idPlaceholder}"></div>`;
                 // Replace test status markers.
                 const plantUmls = text.match(plantUmlRegex) || [];
-                const replacedText = plantUmls.reduce(
-                    (acc: string, plantUml: string, index: number): string =>
-                        acc.replace(new RegExp(escapeStringForUseInRegex(plantUml), 'g'), plantUmlContainer.replace(idPlaceholder, index.toString())),
-                    text,
-                );
+                const replacedText = plantUmls.reduce((acc: string, plantUml: string): string => {
+                    this.plantUmlIndex++;
+                    return acc.replace(new RegExp(escapeStringForUseInRegex(plantUml), 'g'), plantUmlContainer.replace(idPlaceholder, this.plantUmlIndex.toString()));
+                }, text);
                 const plantUmlsValidated = plantUmls.map((plantUml) =>
                     plantUml.replace(/testsColor\(([^)]+)\)/g, (match: any, capture: string) => {
                         const tests = capture.split(',');
                         const { testCaseState } = this.programmingExerciseInstructionService.testStatusForTask(tests, this.latestResult);
                         return testCaseState === TestCaseState.SUCCESS ? 'green' : testCaseState === TestCaseState.FAIL ? 'red' : 'grey';
                     }),
                 );
-                this.injectableElementsFoundSubject.next(() => this.loadAndInjectPlantUmls(plantUmlsValidated));
+                this.injectableElementsFoundSubject.next(() => {
+                    plantUmlsValidated.forEach((plantUml: string, index: number) => {
+                        this.loadAndInjectPlantUml(plantUml, this.plantUmlIndex - plantUmlsValidated.length + index + 1);
```

Comment:
I think this is the correct approach to fixing the issue, however I'm not sure the implementation is bulletproof:

`this.loadAndInjectPlantUml(plantUml, this.plantUmlIndex - plantUmlsValidated.length + index + 1`

As far as I see it, `this.plantUmlIndex - plantUmlsValidated.length` will lead to the wrong id range start for e.g. the second exercise in the exam:

E.g. exercise 1 has 5 umls, exercise 2 has 3.

For exercise 1:
this.plantUmlIndex - plantUmlsValidated.length = 7 - 5 = 2
=> Should be 0

For exercise 2:
this.plantUmlIndex - plantUmlsValidated.length = 7 - 3 = 5
=> Should be 5

I would suggest to directly determine the plantUml ids at the beginning, something like (untested):

```
                        const plantUmls = text.match(plantUmlRegex) || [];
                        // Assign unique ids to uml data structure at the beginning. 
                        const plantUmlsIndexed = plantUmls.map(uml -> {
                            plantUmlIndex++;
                            const nextIndex = plantUmlIndex;
                            return {id: nextIndex, uml}; 
                        })
                        const replacedText = plantUmlsIndexed.reduce((acc: string, {uml, id}): string => {
                            return acc.replace(new RegExp(escapeStringForUseInRegex(uml), 'g'), plantUmlContainer.replace(idPlaceholder, id.toString()));              
                        }, text);
                        ... more stuff about validation, leaving out for brevity sake.
                        // Now here the advantage is that the id is part of the data structure:
                        plantUmlsValidated.forEach(({id, uml}) => {
                            this.loadAndInjectPlantUml(uml, id);
                        }
```

Reply:
Another possible fix would be to save the plantUmlIndex in a closure outside of the thunk.
This has the advantage that the plantUml index can't change.

Again not tested:

```
// Store variable in closure.
const plantUmlIndexRangeStart = this.plantUmlIndex - plantUmlsValidated.length;
plantUmlsValidated.forEach((plantUml: string, index: number) => {]
                        // Access the closure variable instead of the class property.
                        this.loadAndInjectPlantUml(plantUml, plantUmlIndexRangeStart + index + 1);
}
```

Reply:
I assumed that the calculation of the indices is not done at the same time for different exercises (only when you switch between 2). So in your example, it would be instead

For exercise 1:
this.plantUmlIndex - plantUmlsValidated.length = **5** - 5 = 2

However, I agree that my approach is too simple and we should implement this in a more bulletproof fashion. The problem is, I am not so experienced with the mechanisms and did not have enough time to fix that

Reply:
fixed by @kloessst